File: packages/javascript/src/branch-tree.js
const MAX_SEGMENTS = 357_913_941;
const MASK64 = (1n << 64n) - 1n;
const KEYS = ["seed", "root", "rules", "maxSegments"];
const ROOT_KEYS = ["origin", "heading", "length"];
const RULE_KEYS = ["lengthScale", "slots"];
const SLOT_KEYS = ["probability", "turn"];

export class BranchTreeError extends Error {
  constructor(code, parentIndex = undefined, slotIndex = undefined, stage = undefined) {
    super(code); this.name = "BranchTreeError"; this.code = code;
    if (parentIndex !== undefined) this.parentIndex = parentIndex;
    if (slotIndex !== undefined) this.slotIndex = slotIndex;
    if (stage !== undefined) this.stage = stage;
  }
}
function fail(code) { throw new BranchTreeError(code); }
function arithmetic(parentIndex, slotIndex, stage) { throw new BranchTreeError("BRANCH_ARITHMETIC_INVALID", parentIndex, slotIndex, stage); }
function limit(parentIndex, slotIndex) { throw new BranchTreeError("SEGMENT_LIMIT_EXCEEDED", parentIndex, slotIndex); }
function finite(value) { return typeof value === "number" && Number.isFinite(value); }
function zero(value) { return value === 0 ? 0 : value; }
function number(value) { if (!finite(value)) fail("INVALID_INPUT"); return zero(value); }
function passiveRecord(value, keys) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) fail("INVALID_INPUT");
  const prototype = Object.getPrototypeOf(value);
  if (prototype !== Object.prototype && prototype !== null || Reflect.ownKeys(value).length !== keys.length) fail("INVALID_INPUT");
  for (const key of keys) { const descriptor = Object.getOwnPropertyDescriptor(value, key); if (!descriptor || !("value" in descriptor)) fail("INVALID_INPUT"); }
  return value;
}
function value(record, key) { return Object.getOwnPropertyDescriptor(record, key).value; }
function passiveArray(value, length = undefined) {
  if (!Array.isArray(value) || Object.getPrototypeOf(value) !== Array.prototype || (length !== undefined && value.length !== length)) fail("INVALID_INPUT");
  return value;
}
function item(array, index) { const descriptor = Object.getOwnPropertyDescriptor(array, String(index)); if (!descriptor || !("value" in descriptor)) fail("INVALID_INPUT"); return descriptor.value; }
function pair(value) { return passiveArray(value, 2); }
function nonnegative(value) { const result = number(value); if (result < 0) fail("INVALID_INPUT"); return result; }
function integral(value, minimum, maximum) { const result = number(value); if (!Number.isSafeInteger(result) || result < minimum || result > maximum) fail("INVALID_INPUT"); return result; }
function root(value_) {
  const record = passiveRecord(value_, ROOT_KEYS), origin = pair(value(record, "origin"));
  return { x: number(item(origin, 0)), y: number(item(origin, 1)), heading: number(value(record, "heading")), length: nonnegative(value(record, "length")) };
}
function rules(value_) {
  const supplied = passiveArray(value_), output = new Array(supplied.length);
  for (let ruleIndex = 0; ruleIndex < supplied.length; ruleIndex += 1) {
    const record = passiveRecord(item(supplied, ruleIndex), RULE_KEYS), scale = pair(value(record, "lengthScale"));
    const scaleLow = nonnegative(item(scale, 0)), scaleHigh = nonnegative(item(scale, 1)); if (scaleLow > scaleHigh) fail("INVALID_INPUT");
    const suppliedSlots = passiveArray(value(record, "slots")), slots = new Array(suppliedSlots.length);
    for (let slotIndex = 0; slotIndex < suppliedSlots.length; slotIndex += 1) {
      const slot = passiveRecord(item(suppliedSlots, slotIndex), SLOT_KEYS), probability = number(value(slot, "probability"));
      if (probability < 0 || probability > 1) fail("INVALID_INPUT");
      const turn = pair(value(slot, "turn")), low = number(item(turn, 0)), high = number(item(turn, 1)); if (low > high) fail("INVALID_INPUT");
      slots[slotIndex] = { probability, low, high };
    }
    output[ruleIndex] = { scaleLow, scaleHigh, slots };
  }
  return output;
}
function interpolate(low, high, unit) {
  if (unit === 0) return zero(low); if (unit === 1) return zero(high);
  let raw = (low < 0 && high > 0) || (high < 0 && low > 0) ? low * (1 - unit) + high * unit : low + (high - low) * unit;
  if (raw < low) raw = low; else if (raw > high) raw = high;
  return zero(raw);
}
function splitMix(state) { state = ((state ^ (state >> 30n)) * 0xbf58476d1ce4e5b9n) & MASK64; state = ((state ^ (state >> 27n)) * 0x94d049bb133111ebn) & MASK64; return (state ^ (state >> 31n)) & MASK64; }
class Stream {
  constructor(seed) { let state = (BigInt(seed) + 0x9e3779b97f4a7c15n) & MASK64; const first = splitMix(state); state = (state + 0x9e3779b97f4a7c15n) & MASK64; const second = splitMix(state); this.s0 = Number(first & 0xffffffffn) >>> 0; this.s1 = Number(first >> 32n) >>> 0; this.s2 = Number(second & 0xffffffffn) >>> 0; this.s3 = Number(second >> 32n) >>> 0; }
  output() { const value_ = Math.imul(this.s1, 5) >>> 0, result = Math.imul(((value_ << 7) | (value_ >>> 25)) >>> 0, 9) >>> 0, temporary = (this.s1 << 9) >>> 0; this.s2 = (this.s2 ^ this.s0) >>> 0; this.s3 = (this.s3 ^ this.s1) >>> 0; this.s1 = (this.s1 ^ this.s2) >>> 0; this.s0 = (this.s0 ^ this.s3) >>> 0; this.s2 = (this.s2 ^ temporary) >>> 0; this.s3 = ((this.s3 << 11) | (this.s3 >>> 21)) >>> 0; return result; }
  unit() { return this.output() / 4_294_967_296; }
  state() { return [this.s0, this.s1, this.s2, this.s3]; }
}
class Storage {
  constructor(maximum) { this.maximum = maximum; this.size = 0; this.capacity = Math.min(maximum, 16); this.startX = new Float64Array(this.capacity); this.startY = new Float64Array(this.capacity); this.endX = new Float64Array(this.capacity); this.endY = new Float64Array(this.capacity); this.headings = new Float64Array(this.capacity); this.lengths = new Float64Array(this.capacity); this.parents = new Int32Array(this.capacity); this.generations = new Int32Array(this.capacity); this.childCounts = new Int32Array(this.capacity); }
  grow() { if (this.size < this.capacity) return; let capacity = this.capacity; while (capacity < this.size + 1) capacity = Math.min(this.maximum, capacity + Math.max(capacity, 16)); for (const name of ["startX", "startY", "endX", "endY", "headings", "lengths"]) { const next = new Float64Array(capacity); next.set(this[name]); this[name] = next; } for (const name of ["parents", "generations", "childCounts"]) { const next = new Int32Array(capacity); next.set(this[name]); this[name] = next; } this.capacity = capacity; }
  append(x0, y0, x1, y1, heading, length, parent, generation) { this.grow(); const i = this.size++; this.startX[i] = x0; this.startY[i] = y0; this.endX[i] = x1; this.endY[i] = y1; this.headings[i] = heading; this.lengths[i] = length; this.parents[i] = parent; this.generations[i] = generation; this.childCounts[i] = 0; }
  finish() { for (const name of ["startX", "startY", "endX", "endY", "headings", "lengths", "parents", "generations", "childCounts"]) this[name] = this[name].slice(0, this.size); return this; }
}
class Result {
  #storage;
  constructor(storage) { this.#storage = storage; Object.freeze(this); }
  get size() { return this.#storage.size; }
  #index(index) { if (typeof index !== "number" || !Number.isSafeInteger(index) || index < 0) fail("INVALID_INDEX"); if (index >= this.size) fail("INDEX_OUT_OF_RANGE"); return index; }
  segmentAt(index) { const i = this.#index(index), s = this.#storage; return [zero(s.startX[i]), zero(s.startY[i]), zero(s.endX[i]), zero(s.endY[i])]; }
  headingAt(index) { return zero(this.#storage.headings[this.#index(index)]); }
  lengthAt(index) { return zero(this.#storage.lengths[this.#index(index)]); }
  parentAt(index) { return this.#storage.parents[this.#index(index)]; }
  generationAt(index) { return this.#storage.generations[this.#index(index)]; }
  childCountAt(index) { return this.#storage.childCounts[this.#index(index)]; }
  segmentInto(index, out, offset = 0) { const i = this.#index(index); if (!(out instanceof Float64Array) && !(Array.isArray(out) && Object.getPrototypeOf(out) === Array.prototype)) fail("INVALID_OUTPUT"); if (!finite(offset) || !Number.isSafeInteger(offset) || offset < 0 || offset > out.length - 4) fail("INVALID_OUTPUT"); if (Array.isArray(out)) for (let n = 0; n < 4; n += 1) { const descriptor = Object.getOwnPropertyDescriptor(out, String(offset + n)); if (descriptor !== undefined && (!("value" in descriptor) || descriptor.writable !== true) || descriptor === undefined && !Object.isExtensible(out)) fail("INVALID_OUTPUT"); } const s = this.#storage; out[offset] = zero(s.startX[i]); out[offset + 1] = zero(s.startY[i]); out[offset + 2] = zero(s.endX[i]); out[offset + 3] = zero(s.endY[i]); return out; }
  toValues() { const s = this.#storage, result = { segments: new Array(this.size), headings: new Array(this.size), lengths: new Array(this.size), parents: new Array(this.size), generations: new Array(this.size), childCounts: new Array(this.size) }; for (let i = 0; i < this.size; i += 1) { result.segments[i] = this.segmentAt(i); result.headings[i] = zero(s.headings[i]); result.lengths[i] = zero(s.lengths[i]); result.parents[i] = s.parents[i]; result.generations[i] = s.generations[i]; result.childCounts[i] = s.childCounts[i]; } return result; }
}

/**
 * Retain a bounded breadth-first endpoint branch tree.
 * Motivated by survey/out/2018/Generativos/arbolito3/notes.md and
 * survey/out/2018/Generativos/arbolito4/notes.md. Their measured edits establish
 * impact, not defaults or recommended parameter ranges. See
 * topology.seeded-endpoint-branches-2d.
 */
export function seededEndpointBranches2D(config) {
  const record = passiveRecord(config, KEYS), seed = integral(number(value(record, "seed")), 0, 0xffffffff), root_ = root(value(record, "root")), rules_ = rules(value(record, "rules")), maximum = integral(number(value(record, "maxSegments")), 1, MAX_SEGMENTS);
  const rootDx = zero(Math.cos(root_.heading) * root_.length); if (!finite(rootDx)) arithmetic(-1, -1, "delta_x");
  const rootDy = zero(Math.sin(root_.heading) * root_.length); if (!finite(rootDy)) arithmetic(-1, -1, "delta_y");
  const rootEndX = zero(root_.x + rootDx); if (!finite(rootEndX)) arithmetic(-1, -1, "position_x");
  const rootEndY = zero(root_.y + rootDy); if (!finite(rootEndY)) arithmetic(-1, -1, "position_y");
  const storage = new Storage(maximum); storage.append(root_.x, root_.y, rootEndX, rootEndY, root_.heading, root_.length, -1, 0);
  if (rules_.length === 0) return new Result(storage.finish());
  const stream = new Stream(seed);
  for (let parentIndex = 0; parentIndex < storage.size; parentIndex += 1) {
    const generation = storage.generations[parentIndex]; if (generation >= rules_.length) continue;
    const rule = rules_[generation], scale = interpolate(rule.scaleLow, rule.scaleHigh, stream.unit());
    for (let slotIndex = 0; slotIndex < rule.slots.length; slotIndex += 1) { const slot = rule.slots[slotIndex]; if (!(stream.unit() < slot.probability)) continue; if (storage.size >= maximum) limit(parentIndex, slotIndex); const turn = interpolate(slot.low, slot.high, stream.unit()); const childLength = zero(storage.lengths[parentIndex] * scale); if (!finite(childLength)) arithmetic(parentIndex, slotIndex, "length"); const childHeading = zero(storage.headings[parentIndex] + turn); if (!finite(childHeading)) arithmetic(parentIndex, slotIndex, "heading"); const childDx = zero(Math.cos(childHeading) * childLength); if (!finite(childDx)) arithmetic(parentIndex, slotIndex, "delta_x"); const childDy = zero(Math.sin(childHeading) * childLength); if (!finite(childDy)) arithmetic(parentIndex, slotIndex, "delta_y"); const childEndX = zero(storage.endX[parentIndex] + childDx); if (!finite(childEndX)) arithmetic(parentIndex, slotIndex, "position_x"); const childEndY = zero(storage.endY[parentIndex] + childDy); if (!finite(childEndY)) arithmetic(parentIndex, slotIndex, "position_y"); storage.append(storage.endX[parentIndex], storage.endY[parentIndex], childEndX, childEndY, childHeading, childLength, parentIndex, generation + 1); storage.childCounts[parentIndex] += 1; }
  }
  return new Result(storage.finish());
}

File: packages/python/procedurals/branch_tree.py
"""Retained endpoint branch geometry for frozen CP6 contracts."""
from array import array
from dataclasses import dataclass
import math

__all__ = ["BranchTreeError", "seeded_endpoint_branches_2d"]
_MAX_SEGMENTS = 357_913_941
_MAX_SAFE = 9_007_199_254_740_991
_MASK32 = (1 << 32) - 1
_MASK64 = (1 << 64) - 1


class BranchTreeError(ValueError):
    def __init__(self, code, parent_index=None, slot_index=None, stage=None):
        super().__init__(code); self.code = code
        if parent_index is not None: self.parent_index = parent_index
        if slot_index is not None: self.slot_index = slot_index
        if stage is not None: self.stage = stage


def _fail(code="INVALID_INPUT"): raise BranchTreeError(code)
def _dynamic(parent, slot, stage): raise BranchTreeError("BRANCH_ARITHMETIC_INVALID", parent, slot, stage)
def _limit(parent, slot): raise BranchTreeError("SEGMENT_LIMIT_EXCEEDED", parent, slot)
def _zero(value): return 0.0 if value == 0.0 else value
def _number(value):
    if type(value) not in (int, float): _fail()
    try: value = float(value)
    except OverflowError: _fail()
    if not math.isfinite(value): _fail()
    return _zero(value)
def _integer(value, maximum, code="INVALID_INPUT", minimum=0):
    if type(value) is int:
        if value < minimum or value > maximum: _fail(code)
        return value
    if type(value) is float and math.isfinite(value) and value.is_integer() and minimum <= value <= maximum: return int(value)
    _fail(code)
def _record(value, keys):
    if type(value) is not dict or len(value) != len(keys) or any(key not in value for key in keys): _fail()
    return value
def _pair(value):
    if type(value) is not list or len(value) != 2: _fail()
    return value
def _root(value):
    record = _record(value, ("origin", "heading", "length")); origin = _pair(record["origin"])
    x, y, heading, length = _number(origin[0]), _number(origin[1]), _number(record["heading"]), _number(record["length"])
    if length < 0: _fail()
    return x, y, heading, length
def _rules(value):
    if type(value) is not list: _fail()
    output = []
    for raw_rule in value:
        record = _record(raw_rule, ("lengthScale", "slots")); scale = _pair(record["lengthScale"])
        low, high = _number(scale[0]), _number(scale[1])
        if low < 0 or high < 0 or low > high: _fail()
        if type(record["slots"]) is not list: _fail()
        slots = []
        for raw_slot in record["slots"]:
            slot = _record(raw_slot, ("probability", "turn")); probability = _number(slot["probability"])
            if probability < 0 or probability > 1: _fail()
            turn = _pair(slot["turn"])
            turn_low, turn_high = _number(turn[0]), _number(turn[1])
            if turn_low > turn_high: _fail()
            slots.append((probability, turn_low, turn_high))
        output.append((low, high, tuple(slots)))
    return tuple(output)
def _lerp(low, high, unit):
    if unit == 0: return _zero(low)
    if unit == 1: return _zero(high)
    raw = low * (1.0 - unit) + high * unit if (low < 0 < high or high < 0 < low) else low + (high - low) * unit
    if raw < low: raw = low
    elif raw > high: raw = high
    return _zero(raw)


class _Stream:
    def __init__(self, seed):
        state = seed + 0x9E3779B97F4A7C15; first = self._split(state); second = self._split(state + 0x9E3779B97F4A7C15)
        self._state = [first & _MASK32, first >> 32, second & _MASK32, second >> 32]
    @staticmethod
    def _split(value):
        value &= _MASK64; value = ((value ^ (value >> 30)) * 0xBF58476D1CE4E5B9) & _MASK64; value = ((value ^ (value >> 27)) * 0x94D049BB133111EB) & _MASK64
        return (value ^ (value >> 31)) & _MASK64
    def output(self):
        s = self._state; value = (s[1] * 5) & _MASK32; result = ((((value << 7) | (value >> 25)) & _MASK32) * 9) & _MASK32; temporary = (s[1] << 9) & _MASK32
        s[2] ^= s[0]; s[3] ^= s[1]; s[1] ^= s[2]; s[0] ^= s[3]; s[2] ^= temporary; s[3] = ((s[3] << 11) | (s[3] >> 21)) & _MASK32
        return result
    def unit(self): return self.output() / 4294967296.0
    def state(self): return self._state.copy()


@dataclass(frozen=True, slots=True)
class _Result:
    _x0: array; _y0: array; _x1: array; _y1: array; _headings: array; _lengths: array; _parents: array; _generations: array; _children: array
    @property
    def size(self): return len(self._parents)
    def _index(self, index):
        index = _integer(index, _MAX_SAFE, "INVALID_INDEX")
        if index >= self.size: _fail("INDEX_OUT_OF_RANGE")
        return index
    def segment_at(self, index):
        index = self._index(index); return [_zero(self._x0[index]), _zero(self._y0[index]), _zero(self._x1[index]), _zero(self._y1[index])]
    def heading_at(self, index): return _zero(self._headings[self._index(index)])
    def length_at(self, index): return _zero(self._lengths[self._index(index)])
    def parent_at(self, index): return self._parents[self._index(index)]
    def generation_at(self, index): return self._generations[self._index(index)]
    def child_count_at(self, index): return self._children[self._index(index)]
    def segment_into(self, index, out, offset=0):
        index = self._index(index); offset = _integer(offset, _MAX_SAFE, "INVALID_OUTPUT")
        if not (type(out) is list or (type(out) is array and out.typecode == "d" and out.itemsize == 8)) or offset > len(out) - 4: _fail("INVALID_OUTPUT")
        out[offset], out[offset + 1], out[offset + 2], out[offset + 3] = _zero(self._x0[index]), _zero(self._y0[index]), _zero(self._x1[index]), _zero(self._y1[index])
    def to_values(self):
        return {"segments": [self.segment_at(i) for i in range(self.size)], "headings": [self.heading_at(i) for i in range(self.size)], "lengths": [self.length_at(i) for i in range(self.size)], "parents": [self.parent_at(i) for i in range(self.size)], "generations": [self.generation_at(i) for i in range(self.size)], "childCounts": [self.child_count_at(i) for i in range(self.size)]}


def seeded_endpoint_branches_2d(config):
    """Generate retained endpoint branches in breadth-first order.

    Motivation: survey/out/2018/Generativos/arbolito3/notes.md and
    survey/out/2018/Generativos/arbolito4/notes.md. Supply explicit generation
    rules; this operation defines no default or recommended artistic ranges.
    See catalog/operations/seeded-endpoint-branches.json for the frozen contract.
    """
    record = _record(config, ("seed", "root", "rules", "maxSegments")); seed = _integer(record["seed"], _MASK32); x, y, heading, length = _root(record["root"]); rules = _rules(record["rules"]); maximum = _integer(record["maxSegments"], _MAX_SEGMENTS, minimum=1)
    dx = _zero(math.cos(heading) * length)
    if not math.isfinite(dx): _dynamic(-1, -1, "delta_x")
    dy = _zero(math.sin(heading) * length)
    if not math.isfinite(dy): _dynamic(-1, -1, "delta_y")
    end_x = _zero(x + dx)
    if not math.isfinite(end_x): _dynamic(-1, -1, "position_x")
    end_y = _zero(y + dy)
    if not math.isfinite(end_y): _dynamic(-1, -1, "position_y")
    x0, y0, x1, y1, headings, lengths = (array("d") for _ in range(6)); parents, generations, children = (array("i") for _ in range(3))
    def append(a, b, c, d, h, l, parent, generation):
        x0.append(a); y0.append(b); x1.append(c); y1.append(d); headings.append(h); lengths.append(l); parents.append(parent); generations.append(generation); children.append(0)
    append(x, y, end_x, end_y, heading, length, -1, 0)
    if rules:
        stream = _Stream(seed); parent = 0
        while parent < len(parents):
            generation = generations[parent]
            if generation < len(rules):
                low, high, slots = rules[generation]; scale = _lerp(low, high, stream.unit())
                for slot_index, (probability, turn_low, turn_high) in enumerate(slots):
                    if not stream.unit() < probability: continue
                    if len(parents) >= maximum: _limit(parent, slot_index)
                    turn = _lerp(turn_low, turn_high, stream.unit()); child_length = _zero(lengths[parent] * scale)
                    if not math.isfinite(child_length): _dynamic(parent, slot_index, "length")
                    child_heading = _zero(headings[parent] + turn)
                    if not math.isfinite(child_heading): _dynamic(parent, slot_index, "heading")
                    child_dx = _zero(math.cos(child_heading) * child_length)
                    if not math.isfinite(child_dx): _dynamic(parent, slot_index, "delta_x")
                    child_dy = _zero(math.sin(child_heading) * child_length)
                    if not math.isfinite(child_dy): _dynamic(parent, slot_index, "delta_y")
                    child_x = _zero(x1[parent] + child_dx)
                    if not math.isfinite(child_x): _dynamic(parent, slot_index, "position_x")
                    child_y = _zero(y1[parent] + child_dy)
                    if not math.isfinite(child_y): _dynamic(parent, slot_index, "position_y")
                    append(x1[parent], y1[parent], child_x, child_y, child_heading, child_length, parent, generation + 1); children[parent] += 1
            parent += 1
    return _Result(x0, y0, x1, y1, headings, lengths, parents, generations, children)
